rust/debian/Dockerfile

```diff
@@ -34,24 +34,175 @@ RUN --mount=type=secret,id=aws,target=/root/.aws/credentials \
     if [ -z "$SCCACHE_BUCKET" ]; then unset RUSTC_WRAPPER; fi; \
     if [ -z "$SCCACHE_ENDPOINT" ]; then unset SCCACHE_ENDPOINT; fi; \
     mkdir -p /out/tools && \
-    cargo install --git https://github.com/flamegraph-rs/flamegraph \
+    cargo install -q --git https://github.com/flamegraph-rs/flamegraph \
         --tag v0.5.1 --locked --root /out/tools \
         --target=${!TARGETARCH}-unknown-linux-musl && \
-    cargo install cargo-cache --version 0.8.2  --locked --root /out/tools \
+    cargo install -q cargo-cache --version 0.8.2  --locked --root /out/tools \
         --no-default-features --features ci-autoclean,vendored-libgit \
         --target=${!TARGETARCH}-unknown-linux-musl && \
-    cargo install cargo-audit --version 0.16.0  --locked --root /out/tools \
+    cargo install -q cargo-audit --version 0.16.0  --locked --root /out/tools \
         --no-default-features --features vendored-openssl,vendored-libgit2 \
         --target=${!TARGETARCH}-unknown-linux-musl && \
     sccache --show-stats
 
+ENV LINUX_HEADER_VERSION="4.19.88"
+ARG MUSL_VERSION="1.2.2"
+
+ENV LLVM_SYSROOT="/sysroot/${CROSS_COMPILER_TARGET_ARCH}-musl"
+ENV MUSL_TRIPLE="${CROSS_COMPILER_TARGET_ARCH}-unknown-linux-musl"
+
+
+# Create LLVM_SYSROOT
+RUN mkdir -p $LLVM_SYSROOT
+
+# Install linux headers in LLVM_SYSROOT
+RUN cd /tmp && \
+    curl --retry 5 -sSfL -O "https://github.com/sabotage-linux/kernel-headers/archive/v$LINUX_HEADER_VERSION.tar.gz" && \
+    tar xf v$LINUX_HEADER_VERSION.tar.gz && cd kernel-headers-$LINUX_HEADER_VERSION && \
+    MAKEFLAGS="--silent -j $(nproc)" && \
+    make ARCH=${CROSS_COMPILER_TARGET_ARCH} prefix=$LLVM_SYSROOT/usr/ install && \
+    cd / && rm -r /tmp/*
+
+# Download and untar musl
+RUN --mount=type=secret,id=aws,target=/root/.aws/credentials \
+    if [ -z "$SCCACHE_BUCKET" ]; then unset CC_WRAPPER; fi; \
+    if [ -z "$SCCACHE_ENDPOINT" ]; then unset SCCACHE_ENDPOINT; fi; \
+    curl -fSLo /tmp/musl.tar.gz https://www.musl-libc.org/releases/musl-$MUSL_VERSION.tar.gz && \
+    tar -C /tmp -xf /tmp/musl.tar.gz && \
+    cd /tmp/musl-$MUSL_VERSION && \
+    MAKEFLAGS="--silent -j $(nproc)" && \
+    AR=llvm-ar RANLIB=llvm-ranlib CC="$CC_WRAPPER clang" \
+    CFLAGS="--target=$MUSL_TRIPLE" \
+    LDFLAGS="-fuse-ld=lld -L$LLVM_SYSROOT/usr/lib/linux/" \
+    LIBCC="-lclang_rt.builtins-${CROSS_COMPILER_TARGET_ARCH}" \
+    ./configure --prefix=$LLVM_SYSROOT/usr/$MUSL_TRIPLE \
+                --syslibdir=$LLVM_SYSROOT/usr/lib \
+                --build=${!BUILDARCH} \
+                --target=$MUSL_TRIPLE 2>&1 >/dev/null && \
+    make install-headers
+
+# Configure cmake for building llvm headers, compiler-rt and builtins
```

Comment:
Annoyingly llvm/clang doesn't provide a top level target for just building libc++ and it's headers, it wants to build the whole of llvm and clang at once, which we don't want ot waste time doing (since we want to use the official build and don't want to just replace our custom compiled gcc/g++/libc/libstdc++/binutils with another custom toolchain).

Reply:
What percentage of the actual components are we using? Just curious how much time this approach saves.

Reply:
In terms of lines of code, probabbly less than 10%. we don't compile llvm, clang nor any of the supporting binutils whick make up the vast majority of the llvm project

Reply:
Also, compiling clang and llvm takes... a while... They're huge C++ codebases with a lot of templating going on and we really don't want another set of compilers in the images if we can avoid it.

This method lets us compile and install the exact set of components we actually want